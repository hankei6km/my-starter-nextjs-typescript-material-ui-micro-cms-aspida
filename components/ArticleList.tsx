import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Box from '@material-ui/core/Box';
import ArticleItem from './ArticleItem';
import ArticleDetail from './ArticleDetail';
import { SectionArticleIndex } from '../types/pageTypes';
import { pruneClasses } from '../utils/classes';

const useStyles = makeStyles(() => ({
  'ArticleList-root': {
    width: '100%',
    padding: 0
  },
  'ArticleList-detailList': {},
  'ArticleList-itemList': {}
}));
const classNames = [
  'ArticleList-root',
  'ArticleList-detailList',
  'ArticleList-itemList'
];

export type ArticleListComponent = {};

export type ArticleListVariant = {};

type Props = {
  items: SectionArticleIndex[];
  detail?: boolean;
  classes?: { [key: string]: string };
};

const ArticleList = ({ items, detail = true, classes: inClasses }: Props) => {
  const classes = useStyles({ classes: pruneClasses(inClasses, classNames) });
  return (
    <Box
      component={detail ? 'div' : 'ul'}
      className={`${classes['ArticleList-root']} ${
        detail
          ? classes['ArticleList-detailList']
          : classes['ArticleList-itemList']
      }`}
    >
      {detail ? (
        <>
          {items.map((item) => (
            <ArticleDetail
              key={item.id}
              data={item}
              classes={{ ...inClasses }}
            />
          ))}
        </>
      ) : (
        <>
          {items.map((item) => (
            <ArticleItem key={item.id} data={item} classes={{ ...inClasses }} />
          ))}
        </>
      )}
    </Box>
  );
};

export default ArticleList;
