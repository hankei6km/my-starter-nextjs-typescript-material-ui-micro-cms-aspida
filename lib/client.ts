import aspida from '@aspida/fetch';
import api from '../types/client/$api';
import mock from '../types/client/$mock';

export const fetchConfig = (() => {
  const getApiKey = process.env.GET_API_KEY || '';
  if (getApiKey === '' && process.env.NODE_ENV !== 'test') {
    console.error('$GET_API_KEY is not defined.');
  }
  return {
    headers: { 'X-API-KEY': getApiKey }
  };
})();

// aspida の client.xxx の xxx 部分。
// 今回は api.vi[apiName]
// 手動で定義する必要があるので注意:
export const ApiNameArticleValues = ['posts'] as const;
export type ApiNameArticle = typeof ApiNameArticleValues[number];

const clientV1 = (process.env.USE_MOCK_CLIENT_FORCE === 'true' ||
process.env.USE_MOCK_CLIENT === 'true'
  ? mock(aspida(fetch))
  : api(aspida(fetch))
).api.v1;

export default clientV1;
