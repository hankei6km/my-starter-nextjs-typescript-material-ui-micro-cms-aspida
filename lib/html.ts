import unified, { Processor, FrozenProcessor } from 'unified';
import rehypeParse from 'rehype-parse';
import rehypeMinifyWhitespace from 'rehype-minify-whitespace';
import rehypeStringify from 'rehype-stringify';
import rehypeSanitize from 'rehype-sanitize';
import merge from 'deepmerge';
import gh from 'hast-util-sanitize/lib/github.json';
import { Schema } from 'hast-util-sanitize';
import cheerio from 'cheerio';
import parseStyle from 'style-to-object';
// import camelcaseKeys from 'camelcase-keys';  // vedor prefix が jsx styleにならない?
import { Section, SectionContentHtmlChildren } from '../types/pageTypes';

import visit from 'unist-util-visit';
import { Element } from 'hast';
// import { Heading } from 'mdast';
import { Transformer } from 'unified';

const schema = merge(gh, {
  tagNames: ['picture', 'source'],
  attributes: {
    source: ['srcSet', 'sizes'],
    img: ['srcSet', 'sizes', 'className']
  }
});

export function adjustHeading(
  { top }: { top: number } = { top: 4 }
): Transformer {
  // hコンテント HTML と Markdown で本文に入力される見出しは h2 と h3 .
  // h1 は保険で最上位にまるめる、.
  return function transformer(tree): void {
    // 今回は最上位が h3 か h4 に固定
    function visitorH3Top(node: Element, _index: number): void {
      switch (node.tagName) {
        case 'h1':
          node.tagName = 'h3';
          break;
        case 'h2':
          node.tagName = 'h3';
          break;
        case 'h3':
          node.tagName = 'h4';
          break;
      }
    }
    function visitorH4Top(node: Element, _index: number): void {
      switch (node.tagName) {
        case 'h1':
          node.tagName = 'h4';
          break;
        case 'h2':
          node.tagName = 'h4';
          break;
        case 'h3':
          node.tagName = 'h5';
          break;
      }
    }
    visit(tree, 'element', top === 3 ? visitorH3Top : visitorH4Top);
  };
}

export function processorHtml() {
  return unified().use(rehypeParse, { fragment: true });
}

function normalizeProcessor(processor: Processor): FrozenProcessor {
  return processor
    .use(rehypeMinifyWhitespace)
    .use(rehypeSanitize, (schema as unknown) as Schema)
    .use(rehypeStringify)
    .freeze();
}

export function normalizedHtml(processor: Processor, html: string): string {
  let ret = '';
  normalizeProcessor(processor).process(html, (err, file) => {
    if (err) {
      console.error(err);
    }
    ret = String(file);
  });
  return ret;
}

export function styleToJsxStyle(
  s: string
): SectionContentHtmlChildren['style'] {
  const p = parseStyle(s) || {};
  const ret: SectionContentHtmlChildren['style'] = {};
  Object.entries(p).forEach((kv) => {
    const [first, ...names] = kv[0].split('-');
    const capitalizedNames = names.map(
      (r) => `${r[0].toLocaleUpperCase()}${r.slice(1)}`
    );
    const k = `${first}${capitalizedNames.join('')}`;

    ret[k] = kv[1];
  });
  return ret;
}

// とりあえず、普通に記述された markdown から変換されたときに body の直下にありそうなタグ.
// いまのところ小文字のみ.
export const SectionContentHtmlChildrenElemValues = [
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'div',
  'p',
  'blockquote',
  'code',
  'pre',
  'span',
  'img',
  'picture',
  'table',
  'ul',
  'ol',
  'dl',
  'hr',
  'a',
  'header',
  'footer',
  'section',
  'article',
  'aside'
] as const;
export function htmlToChildren(html: string): SectionContentHtmlChildren[] {
  // markdown の変換と sanitize には unifed を使っているので、少し迷ったが cheerio を利用.
  // (unified は typescript で利用するときに対応していないプラグインがたまにあるので)

  const ret: SectionContentHtmlChildren[] = [];
  const $ = cheerio.load(html);
  try {
    $('body')
      .contents()
      .each((_idx, elm) => {
        // https://stackoverflow.com/questions/31949521/scraping-text-with-cheerio
        // if (elm.nodeType === Node.ELEMENT_NODE) {
        if (elm.nodeType === 1) {
          if (
            SectionContentHtmlChildrenElemValues.some((v) => v === elm.tagName)
          ) {
            const $elm = $(elm);
            const attribs = elm.attribs ? { ...elm.attribs } : {};
            let style: SectionContentHtmlChildren['style'] = {};
            if (elm.attribs.style) {
              style = styleToJsxStyle(elm.attribs.style);
              delete attribs.style;
            }
            if (elm.attribs.class) {
              attribs.className = elm.attribs.class;
              delete attribs.class;
            }
            ret.push({
              tagName: elm.tagName,
              style: style,
              attribs: attribs,
              html: $elm.html() || ''
            });
          } else {
            throw new Error(
              `support only ${SectionContentHtmlChildrenElemValues.join(', ')}`
            );
          }
        } else {
          throw new Error('support only nodeType=Node.ELEMENT_NODE');
        }
      });
  } catch (_e) {
    ret.splice(0, ret.length, {
      tagName: 'div',
      style: {},
      attribs: {},
      html: html
    });
  }
  if (ret.length === 0) {
    ret.push({
      tagName: 'div',
      style: {},
      attribs: {},
      html: html
    });
  }
  return ret;
}

export type IndexedHtml = {
  index: {
    range: [
      // start and end pos.
      number,
      number
    ];
    // section  section contentHtml を特定する idx.
    sectionIdx: number;
    contentIdx: number;
    childIdx: number;
  }[];
  html: string;
};

export function getIndexedHtml(sections: Section[]): IndexedHtml {
  const ret: IndexedHtml = {
    index: [],
    html: ''
  };
  sections.forEach((section, sectionIdx) => {
    section.content.forEach((content, contentIdx) => {
      if (content.kind === 'html') {
        // filter ではじくと index がずれる
        // const contentHtmlWithoutCodeBlock = content.contentHtml.filter(
        //   ({ tagName, html }) =>
        //     tagName !== 'pre' && html.slice(0, 4) !== 'code'
        // );
        content.contentHtml.forEach(({ tagName, html }, childIdx) => {
          if (
            tagName !== 'img' &&
            !(tagName === 'pre' && html.slice(0, 5) === '<code')
          ) {
            const start = ret.html.length;
            ret.html = `${ret.html}<${tagName}>${html}</${tagName}>`;
            const index: IndexedHtml['index'][0] = {
              range: [start, ret.html.length - 1],
              sectionIdx: sectionIdx,
              contentIdx: contentIdx,
              childIdx: childIdx
            };
            ret.index.push(index);
          }
        });
      }
    });
  });
  return ret;
}
