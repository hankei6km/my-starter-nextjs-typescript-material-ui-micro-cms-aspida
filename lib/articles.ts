import { ParsedUrlQuery } from 'querystring';
import { GetStaticPropsContext } from 'next';
import client, { fetchConfig, ApiNameArticle } from './client';
import { ArticlePost, blankArticlePost } from '../types/articleTypes';
import {
  ArticleContent,
  blankArticleContent
} from '../types/client/contentTypes';
import { markdownToHtml } from './markdown';

export async function getSortedArticleList(apiName: ApiNameArticle) {
  try {
    const res = await client[apiName].get({
      query: {
        fields: 'id,createdAt,updatedAt,publishedAt,revisedAt,title'
      },
      config: fetchConfig
    });
    return res.body.contents;
  } catch (err) {
    // res.status = 404 などでも throw される(試した限りでは)
    // res.status を知る方法は?
    console.error(`getSortedArticleList error: ${apiName}: ${err.name}`);
  }
  return [];
}

export async function getAllArticleIds(apiName: ApiNameArticle) {
  try {
    const res = await client[apiName].get({
      query: {
        fields: 'id'
      },
      config: fetchConfig
    });
    return res.body.contents.map(({ id }) => ({ params: { id } }));
  } catch (err) {
    console.error(`getAllArticleIds error: ${apiName}: ${err.name}`);
  }
  return [];
}

export async function getArticleData(
  apiName: ApiNameArticle,
  {
    params = { id: '' },
    preview = false,
    previewData = {}
  }: GetStaticPropsContext<ParsedUrlQuery>
): Promise<ArticleContent> {
  try {
    const res = await client[apiName]
      ._id(!preview ? params.id : previewData.slug) // 似たような3項式がバラけていてすっきりしない
      .$get({
        query: {
          draftKey: !preview ? '' : previewData.draftKey
        },
        config: fetchConfig
      });
    return res;
  } catch (err) {
    console.error(`getArticleData error: ${apiName}: ${err.name}`);
  }
  return blankArticleContent();
}

export async function getArticlePostData(
  apiName: ApiNameArticle,
  {
    params = { id: '' },
    preview = false,
    previewData = {}
  }: GetStaticPropsContext<ParsedUrlQuery>
): Promise<ArticlePost> {
  try {
    const res = await getArticleData(apiName, {
      params,
      preview,
      previewData
    });
    const data = {
      id: res.id,
      updated: res.revisedAt || res.updatedAt,
      title: res.title,
      content: res.content.map((content) => {
        if (content.fieldId === 'contentHtml') {
          return { kind: 'html' as 'html', html: content.html };
        } else if (content.fieldId === 'contentMarkdown') {
          return {
            kind: 'html' as 'html',
            html: markdownToHtml(content.markdown)
          };
        }
        return { kind: 'html' as 'html', html: '' };
      }),
      mainImage: res.mainImage || ''
    };
    return data;
  } catch (err) {
    console.error(`getArticlePostData error: ${apiName}: ${err.name}`);
  }
  return blankArticlePost();
}
