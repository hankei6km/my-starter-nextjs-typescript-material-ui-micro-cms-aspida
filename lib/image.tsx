import ReactDomServer from 'react-dom/server';
import { ImageTemplate, imageTransformedUrl } from './imageTemplate';
import { normalizedHtml, processorHtml } from './html';
import siteServerSideConfig from '../src/site.server-side-config';

export type ImageInfo = {
  // カラーパレット、顔認識情報等も含める予定
  url: string;
  width: number;
  height: number;
};
export type ContentImage = {
  image: ImageInfo;
  alt: string;
  link?: string;
  newTab?: boolean;
  asThumb?: boolean;
};
function toElm(
  contentImage: ContentImage,
  className: string,
  templateList: ImageTemplate[]
): string {
  const idx = templateList.findIndex((t) => t.matcher(contentImage));
  if (idx >= 0) {
    const intermediate = templateList[idx].intermediate;
    if (intermediate.kind === 'picture') {
      // picture タグのみ.
      // descriptor 'x' で決め打ち('w' でも 'x' 扱い)
      const PictureTag = (
        <picture>
          {intermediate.sources.map((source, i) => {
            return (
              <source
                key={i}
                media={source.suggestMedia}
                srcSet={source.srcset.set
                  .map((set) => {
                    return `${contentImage.image.url}?${set.src.url.paramsStr} ${set.density}${source.srcset.descriptor}`;
                  })
                  .join(' ,')}
              />
            );
          })}
          <img
            src={`${contentImage.image.url}?${intermediate.img.src.url.paramsStr}`}
            alt={contentImage.alt}
            className={className}
            width={intermediate.img.width}
            height={intermediate.img.height}
          />
        </picture>
      );
      if (contentImage.asThumb) {
        const largeImage = templateList[idx].largeImage;
        const largeImageUrl = largeImage
          ? imageTransformedUrl(contentImage, largeImage)
          : contentImage.image.url;
        return ReactDomServer.renderToStaticMarkup(
          <a href={largeImageUrl} target="_blank" rel="noopener noreferrer">
            {PictureTag}
          </a>
        );
      }
      return ReactDomServer.renderToStaticMarkup(PictureTag);
    }
    return '';
  }
  // とりあえず.
  const width =
    contentImage.image.width > contentImage.image.height
      ? 500
      : Math.round(
          (contentImage.image.width * 500) / contentImage.image.height
        );
  const height =
    contentImage.image.width > contentImage.image.height
      ? Math.round((contentImage.image.height * 500) / contentImage.image.width)
      : 500;
  const q = new URLSearchParams('');
  q.append('auto', 'compress');
  q.append('w', `${width}`);
  q.append('h', `${height}`);
  const imgTag = (
    <img
      src={`${contentImage.image.url}?${q.toString()}`}
      alt={contentImage.alt}
      // style={{ maxWidth: '100%' }}
      className={className}
      width={width}
      height={height}
    />
  );
  if (contentImage.asThumb) {
    return ReactDomServer.renderToStaticMarkup(
      <a
        href={contentImage.image.url}
        target="_blank"
        rel="noopener noreferrer"
      >
        {imgTag}
      </a>
    );
  }
  return ReactDomServer.renderToStaticMarkup(imgTag);
}

export function imageToHtml(contentImage: ContentImage): string {
  const image = toElm(
    contentImage,
    siteServerSideConfig.imageConfig.contentImageClassName,
    siteServerSideConfig.imageConfig.template
  );
  // normalizeHtml は section の組み立てでコールしたいが、
  // 何か影響ある?
  return normalizedHtml(processorHtml(), image);
}

const fmJsonQuery = (() => {
  const q = new URLSearchParams('');
  q.append('fm', 'json');
  return q.toString();
})();

export async function imageInfo(imageUrl: string): Promise<ImageInfo> {
  const res = await fetch(`${imageUrl}?${fmJsonQuery}`).catch((err) => {
    console.error(err);
    throw new Error(err);
  });
  const info = await res.json();
  const imageInfo: ImageInfo = {
    url: imageUrl,
    width: 0,
    height: 0
  };
  // この辺の詳細な仕様は不明なので様子見.
  // Orientationについて
  // イメージセンサーからのビットマップデータの保存方向(軸)とカメラの向きの関係性?
  // 背面カメラ
  // スマホを通常の持ち方にする(縦長) : 6
  // スマホを左に倒す(横長) : 1
  // スマホを右に倒す(横長) : 3
  // スマホをさかさまにする(縦長) : 8
  // 前面カメラ
  // スマホを通常の持ち方にする(縦長) : 8
  // スマホを左に倒す(背面カメラの向きに直すと右に倒す)(横長) : 3
  // あとの２方向は省略.
  // ここで必要なのは縦長か横長かの情報なので、
  // 以下のよれば 6 以上なら縦横を入れ替えた方がよさそう
  // https://github.com/blueimp/JavaScript-Load-Image/blob/master/js/load-image-orientation.js
  if (info.TIFF && info.Orientation && info.Orientation >= 6) {
    // アスペクト比的なものも保存しておく?
    imageInfo.width = info.PixelHeight;
    imageInfo.height = info.PixelWidth;
  } else {
    imageInfo.width = info.PixelWidth;
    imageInfo.height = info.PixelHeight;
  }

  return imageInfo;
}
