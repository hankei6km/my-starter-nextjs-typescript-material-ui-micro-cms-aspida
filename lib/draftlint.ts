import cheerio from 'cheerio';
import { TextlintKernel } from '@textlint/kernel';
import { TextlintKernelOptions } from '@textlint/kernel/lib/textlint-kernel-interface';
import { Section } from '../types/pageTypes';
import { getIndexedHtml } from './html';
import { getTextlintKernelOptions } from '../utils/textlint';
import traverse from 'traverse';
const parseHtml = require('textlint-plugin-html/lib/html-to-ast').parse;

type InsInfo = {
  message: string;
  ruleId: string;
  severity: number;
  index: number;
  insIndex: number;
  range: [number, number];
};
type InsInfos = InsInfo[];

export function getInsInfos(html: string, messages: any[]): InsInfos {
  const ret: InsInfos = [];
  const nodes = traverse(parseHtml(html)).nodes();
  messages.forEach((m) => {
    let hit = -1;
    const index = m.index;
    nodes.forEach((node, ii) => {
      if (
        node.type &&
        node.range &&
        node.range[0] <= index &&
        index <= node.range[1]
      ) {
        hit = ii;
      }
    });
    if (hit >= 0) {
      const node = nodes[hit];
      const end = m.index - node.range[0];
      // console.log(end, node.value.slice(0, end), node.raw.slice(0, end));
      ret.push({
        message: m.message,
        ruleId: m.ruleId,
        severity: m.severity,
        index: m.index,
        // 末尾 ';' だと欠ける? よって先頭を使う
        // TODO: 末尾の扱いチェック(どこで欠ける?).
        insIndex:
          node.value &&
          node.raw &&
          node.value.slice(0, end) === node.raw.slice(0, end)
            ? m.index
            : node.range[0] - 1,
        range: [node.range[0] as number, node.range[1] as number]
      });
    }
  });
  return ret.sort((a, b) => a.insIndex - b.insIndex);
}

export function insertHtmlToSections(
  html: string,
  pos: number, // section を IndexedHtml の html に変換した場合の html の位置
  sections: Section[]
): Section[] {
  //  const ret = [...sections];
  const { index } = getIndexedHtml(sections);
  const idx = index.findIndex(
    ({ range }) => range[0] <= pos && pos <= range[1]
  );
  if (idx >= 0) {
    const { range, sectionIdx, contentIdx, childIdx } = index[idx];
    const content = sections[sectionIdx].content[contentIdx];
    if (content.kind === 'html') {
      const c = content.contentHtml[childIdx];
      let posInChild = pos - range[0] - (c.tagName.length + 2); // 2 = '<>'.length
      posInChild = posInChild < 0 ? 0 : posInChild;
      c.html = `${c.html.slice(0, posInChild)}${html}${c.html.slice(
        posInChild
      )}`;
    }
  } // 見つからないときは?
  return sections;
}
type TextLintInSectionsResult = {
  sections: Section[];
  messages: { ruleId: string; message: string; id: string; severity: number }[];
  list: string;
};

export async function textLintInSections(
  sections: Section[],
  options?: TextlintKernelOptions,
  messageStyle: { [key: string]: string } = { color: 'red' },
  idPrefix: string = ''
): Promise<TextLintInSectionsResult> {
  let ret: TextLintInSectionsResult = { sections, messages: [], list: '' };
  const indexedHtml = getIndexedHtml(sections);

  // https://github.com/mobilusoss/textlint-browser-runner/tree/master/packages/textlint-bundler
  const kernel = new TextlintKernel();

  const results = [
    await kernel.lintText(
      indexedHtml.html,
      // todo: options(presets など)は SiteServerSideConfig で定義できるようにする.
      options || getTextlintKernelOptions()
    )
  ];
  if (results && results.length > 0 && results[0].messages.length > 0) {
    const insInfos = getInsInfos(indexedHtml.html, results[0].messages);
    let slider = 0;
    const $wrapper = cheerio.load('<span/>')('span');
    Object.entries(messageStyle).forEach(([k, v]) => {
      $wrapper.css(k, v);
    });
    insInfos.forEach((m, i) => {
      const id = `${idPrefix}:textLintMessage:${i}`;
      $wrapper.attr('id', id);
      const html = $wrapper.html(m.message).parent().html();
      if (m.message && html) {
        const index = m.insIndex + 1;
        ret.sections = insertHtmlToSections(html, index + slider, ret.sections);
        slider = slider + html.length;
        ret.messages.push({
          ruleId: m.ruleId,
          message: m.message,
          id,
          severity: m.severity
        });
      }
    });
    const $dl = cheerio.load('<dl/>')('dl');
    ret.messages.forEach((m) => {
      const $d = cheerio.load('<dt></dt><dd><a/></dd>');
      // https://github.com/textlint/textlint/blob/master/packages/%40textlint/kernel/src/shared/rule-severity.ts
      // https://github.com/textlint/textlint/blob/master/packages/%40textlint/kernel/src/context/TextlintRuleSeverityLevelKeys.ts
      $d('dt').text(
        m.severity === 0 ? 'info' : m.severity === 1 ? 'warning' : 'error'
      );
      const $a = $d('a');
      $a.attr('href', `#${m.id}`);
      $a.text(`${m.message}(${m.ruleId})`);
      $dl.append($d('body').children());
    });
    ret.list = $dl.parent().html() || '';
  }
  return ret;
}
