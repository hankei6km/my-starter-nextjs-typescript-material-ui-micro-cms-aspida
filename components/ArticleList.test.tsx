import React from 'react';
import { render, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { RouterContext } from 'next/dist/next-server/lib/router-context';
import {
  mockRouter,
  mockImage,
  mockDataSectionArticleList
} from '../test/testUtils';

import ArticleList from './ArticleList';

// https://stackoverflow.com/questions/40449434/mocking-globals-in-jest
const orgImage = global.Image;
afterEach(() => {
  global.Image = orgImage;
});

describe('ArticleList', () => {
  it('should renders ArticleList', async () => {
    const addEventListener = jest.fn();
    global.Image = mockImage(jest.fn(), addEventListener);

    await act(async () => {
      const router = mockRouter();
      const { container } = render(
        <RouterContext.Provider value={router}>
          <ArticleList items={mockDataSectionArticleList} />
        </RouterContext.Provider>
      );

      await new Promise<void>((resolve) => {
        setTimeout(() => {
          addEventListener.mock.calls.forEach((v) => {
            v[1]();
          });
          resolve();
        }, 10);
      });
      expect(container).toMatchSnapshot();
    });
  });
});
