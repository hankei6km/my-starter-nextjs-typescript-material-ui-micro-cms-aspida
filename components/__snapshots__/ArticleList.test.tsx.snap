// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleList should renders ArticleList 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-5 makeStyles-ArticleList-root-1"
  >
    <div
      class="MuiBox-root MuiBox-root-6"
    >
      <div
        class="MuiBox-root MuiBox-root-7 makeStyles-ArticleList-itemDetail-3"
      >
        <div
          class="MuiPaper-root MuiCard-root makeStyles-ArticleDetail-root-8 MuiPaper-elevation0 MuiPaper-rounded"
        >
          <article
            class="MuiCardContent-root makeStyles-ArticleDetail-content-9"
          >
            <div
              class="MuiCardHeader-root makeStyles-ArticleDetail-title-10"
            >
              <div
                class="MuiCardHeader-content"
              >
                <span
                  class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
                >
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                    href="/test1/zzzzzzzzz"
                  >
                    <h3
                      class="MuiTypography-root MuiTypography-h3 MuiTypography-colorTextPrimary"
                    >
                      title3
                    </h3>
                  </a>
                </span>
                <span
                  class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                >
                  <div
                    class="MuiBox-root MuiBox-root-14 makeStyles-root-13"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorAction MuiSvgIcon-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                      />
                      <path
                        d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
                      />
                    </svg>
                    <div
                      class="MuiTypography-root MuiTypography-body2"
                    >
                      <time
                        datetime="2020-12-27T04:04:30.107Z"
                      >
                        2020-12-27
                      </time>
                    </div>
                  </div>
                </span>
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-15 makeStyles-ArticleDetail-thumbImage-11"
            >
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                href="/test1/zzzzzzzzz"
              >
                <div
                  class="makeStyles-thumb-16"
                >
                  <img
                    alt="thumbnail for title3"
                    height="150"
                    width="250"
                  />
                </div>
              </a>
            </div>
          </article>
          <div
            class="MuiCardActions-root makeStyles-ArticleDetail-actions-12 MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-outlinedPrimary MuiButton-textSizeSmall MuiButton-sizeSmall MuiTypography-colorPrimary"
              href="/test1/zzzzzzzzz"
              role="button"
              style="text-transform: none;"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                read more
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-17"
    >
      <div
        class="MuiBox-root MuiBox-root-18 makeStyles-ArticleList-itemDetail-3"
      >
        <div
          class="MuiPaper-root MuiCard-root makeStyles-ArticleDetail-root-8 MuiPaper-elevation0 MuiPaper-rounded"
        >
          <article
            class="MuiCardContent-root makeStyles-ArticleDetail-content-9"
          >
            <div
              class="MuiCardHeader-root makeStyles-ArticleDetail-title-10"
            >
              <div
                class="MuiCardHeader-content"
              >
                <span
                  class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
                >
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                    href="/test1/yyyyyy-da"
                  >
                    <h3
                      class="MuiTypography-root MuiTypography-h3 MuiTypography-colorTextPrimary"
                    >
                      title2
                    </h3>
                  </a>
                </span>
                <span
                  class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                >
                  <div
                    class="MuiBox-root MuiBox-root-19 makeStyles-root-13"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorAction MuiSvgIcon-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                      />
                      <path
                        d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
                      />
                    </svg>
                    <div
                      class="MuiTypography-root MuiTypography-body2"
                    >
                      <time
                        datetime="2020-12-26T15:29:14.476Z"
                      >
                        2020-12-26
                      </time>
                    </div>
                  </div>
                </span>
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-20 makeStyles-ArticleDetail-thumbImage-11"
            >
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                href="/test1/yyyyyy-da"
              >
                <div
                  class="makeStyles-thumb-16"
                >
                  <img
                    alt="thumbnail for title2"
                    height="150"
                    width="250"
                  />
                </div>
              </a>
            </div>
          </article>
          <div
            class="MuiCardActions-root makeStyles-ArticleDetail-actions-12 MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-outlinedPrimary MuiButton-textSizeSmall MuiButton-sizeSmall MuiTypography-colorPrimary"
              href="/test1/yyyyyy-da"
              role="button"
              style="text-transform: none;"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                read more
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-21"
    >
      <div
        class="MuiBox-root MuiBox-root-22 makeStyles-ArticleList-itemDetail-3"
      >
        <div
          class="MuiPaper-root MuiCard-root makeStyles-ArticleDetail-root-8 MuiPaper-elevation0 MuiPaper-rounded"
        >
          <article
            class="MuiCardContent-root makeStyles-ArticleDetail-content-9"
          >
            <div
              class="MuiCardHeader-root makeStyles-ArticleDetail-title-10"
            >
              <div
                class="MuiCardHeader-content"
              >
                <span
                  class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
                >
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                    href="/test1/xxxxxxxxx"
                  >
                    <h3
                      class="MuiTypography-root MuiTypography-h3 MuiTypography-colorTextPrimary"
                    >
                      title1
                    </h3>
                  </a>
                </span>
                <span
                  class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                >
                  <div
                    class="MuiBox-root MuiBox-root-23 makeStyles-root-13"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-colorAction MuiSvgIcon-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                      />
                      <path
                        d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
                      />
                    </svg>
                    <div
                      class="MuiTypography-root MuiTypography-body2"
                    >
                      <time
                        datetime="2020-12-26T12:27:22.533Z"
                      >
                        2020-12-26
                      </time>
                    </div>
                  </div>
                </span>
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-24 makeStyles-ArticleDetail-thumbImage-11"
            >
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                href="/test1/xxxxxxxxx"
              >
                <div
                  class="makeStyles-thumb-16"
                >
                  <img
                    alt="thumbnail for title1"
                    height="150"
                    width="250"
                  />
                </div>
              </a>
            </div>
          </article>
          <div
            class="MuiCardActions-root makeStyles-ArticleDetail-actions-12 MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-outlinedPrimary MuiButton-textSizeSmall MuiButton-sizeSmall MuiTypography-colorPrimary"
              href="/test1/xxxxxxxxx"
              role="button"
              style="text-transform: none;"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                read more
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
